core/distributed/partition_helpers.cpp

```diff
@@ -0,0 +1,153 @@
+GKO_REGISTER_OPERATION(compress_ranges, partition_helpers::compress_ranges);
+
+
+}  // namespace
+}  // namespace partition_helpers
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+std::unique_ptr<Partition<LocalIndexType, GlobalIndexType>>
+build_partition_from_local_range(std::shared_ptr<const Executor> exec,
+                                 mpi::communicator comm, span local_range)
+{
+    std::array<GlobalIndexType, 2> range{
+        static_cast<GlobalIndexType>(local_range.begin),
+        static_cast<GlobalIndexType>(local_range.end)};
+
+    // make all range_start_ends available on each rank
+    // note: not all combination of MPI + GPU library seem to support
+    // mixing host and device buffers, e.g. OpenMPI 4.0.5 and Rocm 4.0
+    auto mpi_exec = exec->get_master();
+    array<GlobalIndexType> ranges_start_end(mpi_exec, comm.size() * 2);
+    ranges_start_end.fill(invalid_index<GlobalIndexType>());
+    comm.all_gather(mpi_exec, range.data(), 2, ranges_start_end.get_data(), 2);
+    ranges_start_end.set_executor(exec);
+
+    // make_sort_by_range_start
+    array<comm_index_type> part_ids(exec, comm.size());
+    exec->run(components::make_fill_seq_array(part_ids.get_data(),
+                                              part_ids.get_num_elems()));
+    exec->run(partition_helpers::make_sort_by_range_start(ranges_start_end,
+                                                          part_ids));
+
+    // check for consistency
+    bool consecutive_ranges = false;
+    exec->run(partition_helpers::make_check_consecutive_ranges(
+        ranges_start_end, &consecutive_ranges));
+    if (!consecutive_ranges) {
+        throw Error(__FILE__, __LINE__, "The partition contains gaps.");
+    }
+
+    // join (now consecutive) starts and ends into combined array
+    array<GlobalIndexType> ranges(exec, comm.size() + 1);
+    exec->run(
+        partition_helpers::make_compress_ranges(ranges_start_end, ranges));
+
+    return Partition<LocalIndexType, GlobalIndexType>::build_from_contiguous(
+        exec, ranges, part_ids);
+}
+
+#define GKO_DECLARE_BUILD_PARTITION_FROM_LOCAL_RANGE(_local_type,          \
+                                                     _global_type)         \
+    std::unique_ptr<Partition<_local_type, _global_type>>                  \
+    build_partition_from_local_range(std::shared_ptr<const Executor> exec, \
+                                     mpi::communicator comm, span local_range)
+GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
+    GKO_DECLARE_BUILD_PARTITION_FROM_LOCAL_RANGE);
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+std::unique_ptr<Partition<LocalIndexType, GlobalIndexType>>
+build_partition_from_local_size(std::shared_ptr<const Executor> exec,
+                                mpi::communicator comm, size_type local_size)
+{
+    auto local_size_gi = static_cast<GlobalIndexType>(local_size);
+    std::vector<GlobalIndexType> sizes(comm.size());
```

Comment:
I agree that it is more comfortable, but we avoid our loggers with `std::vector`. And since you don't need resizing or other more advanced features of `std::vector`, using `gko::array` seems more logical to me.